This notebook demonstrates advanced SQL techniques like CASE statements, subqueries, WITH clauses, and window functions through analysis of the European Soccer Database. The soccer database contains team, match, and player data for top European leagues from 2008-2016.

After importing necessary libraries like pandas and sqlite3, SQL queries are constructed to explore the database. Techniques like conditional logic in CASE statements, nesting subqueries, common table expressions with WITH clause, and analytic functions like ranking are applied.

The goals are to 
- gain insights into the soccer data, and 
- exemplify effective use of advanced SQL constructs for performing complex analysis. The notebook serves both as a practical guide for learning SQL and a showcase of impactful techniques using a real-world sports dataset.

In [1]:
import numpy as np
import pandas as pd
import sqlite3
import matplotlib.pyplot as plt

conn = sqlite3.connect('/kaggle/input/soccer/database.sqlite')

In [2]:
tables = pd.read_sql("""SELECT *
                        FROM sqlite_master
                        WHERE type='table';""", conn)
tables


type               name           tbl_name  rootpage  \
0  table    sqlite_sequence    sqlite_sequence         4   
1  table  Player_Attributes  Player_Attributes        11   
2  table             Player             Player        14   
3  table              Match              Match        18   
4  table             League             League        24   
5  table            Country            Country        26   
6  table               Team               Team        29   
7  table    Team_Attributes    Team_Attributes         2   

                                                 sql  
0             CREATE TABLE sqlite_sequence(name,seq)  
1  CREATE TABLE "Player_Attributes" (\n\t`id`\tIN...  
2  CREATE TABLE `Player` (\n\t`id`\tINTEGER PRIMA...  
3  CREATE TABLE `Match` (\n\t`id`\tINTEGER PRIMAR...  
4  CREATE TABLE `League` (\n\t`id`\tINTEGER PRIMA...  
5  CREATE TABLE `Country` (\n\t`id`\tINTEGER PRIM...  
6  CREATE TABLE "Team" (\n\t`id`\tINTEGER PRIMARY...  
7  CREATE TABLE `Team_Attributes` (\n\t`id`\tINTE...

In [3]:
match_table = pd.read_sql("""SELECT *
                        FROM Match;""", conn)
match_table


id  country_id  league_id     season  stage                 date  \
0          1           1          1  2008/2009      1  2008-08-17 00:00:00   
1          2           1          1  2008/2009      1  2008-08-16 00:00:00   
2          3           1          1  2008/2009      1  2008-08-16 00:00:00   
3          4           1          1  2008/2009      1  2008-08-17 00:00:00   
4          5           1          1  2008/2009      1  2008-08-16 00:00:00   
...      ...         ...        ...        ...    ...                  ...   
25974  25975       24558      24558  2015/2016      9  2015-09-22 00:00:00   
25975  25976       24558      24558  2015/2016      9  2015-09-23 00:00:00   
25976  25977       24558      24558  2015/2016      9  2015-09-23 00:00:00   
25977  25978       24558      24558  2015/2016      9  2015-09-22 00:00:00   
25978  25979       24558      24558  2015/2016      9  2015-09-23 00:00:00   

       match_api_id  home_team_api_id  away_team_api_id  home_team_goal  ...  \
0            492473              9987              9993               1  ...   
1            492474             10000              9994               0  ...   
2            492475              9984              8635               0  ...   
3            492476              9991              9998               5  ...   
4            492477              7947              9985               1  ...   
...             ...               ...               ...             ...  ...   
25974       1992091             10190             10191               1  ...   
25975       1992092              9824             10199               1  ...   
25976       1992093              9956             10179               2  ...   
25977       1992094              7896             10243               0  ...   
25978       1992095             10192              9931               4  ...   

        SJA   VCH   VCD   VCA   GBH   GBD   GBA   BSH   BSD   BSA  
0      4.00  1.65  3.40  4.50  1.78  3.25  4.00  1.73  3.40  4.20  
1      3.80  2.00  3.25  3.25  1.85  3.25  3.75  1.91  3.25  3.60  
2      2.50  2.35  3.25  2.65  2.50  3.20  2.50  2.30  3.20  2.75  
3      7.50  1.45  3.75  6.50  1.50  3.75  5.50  1.44  3.75  6.50  
4      1.73  4.50  3.40  1.65  4.50  3.50  1.65  4.75  3.30  1.67  
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  
25974   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
25975   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
25976   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
25977   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
25978   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  

[25979 rows x 115 columns]

In [4]:
Player_table = pd.read_sql("""SELECT *
                        FROM Player;""", conn)
Player_table

id  player_api_id          player_name  player_fifa_api_id  \
0          1         505942   Aaron Appindangoye              218353   
1          2         155782      Aaron Cresswell              189615   
2          3         162549          Aaron Doran              186170   
3          4          30572        Aaron Galindo              140161   
4          5          23780         Aaron Hughes               17725   
...      ...            ...                  ...                 ...   
11055  11071          26357       Zoumana Camara                2488   
11056  11072         111182         Zsolt Laczko              164680   
11057  11073          36491            Zsolt Low              111191   
11058  11074          35506  Zurab Khizanishvili               47058   
11059  11075          39902   Zvjezdan Misimovic              102359   

                  birthday  height  weight  
0      1992-02-29 00:00:00  182.88     187  
1      1989-12-15 00:00:00  170.18     146  
2      1991-05-13 00:00:00  170.18     163  
3      1982-05-08 00:00:00  182.88     198  
4      1979-11-08 00:00:00  182.88     154  
...                    ...     ...     ...  
11055  1979-04-03 00:00:00  182.88     168  
11056  1986-12-18 00:00:00  182.88     176  
11057  1979-04-29 00:00:00  180.34     154  
11058  1981-10-06 00:00:00  185.42     172  
11059  1982-06-05 00:00:00  180.34     176  

[11060 rows x 7 columns]

In [5]:
League_table = pd.read_sql("""SELECT *
                        FROM League;""", conn)
League_table

id  country_id                      name
0       1           1    Belgium Jupiler League
1    1729        1729    England Premier League
2    4769        4769            France Ligue 1
3    7809        7809     Germany 1. Bundesliga
4   10257       10257             Italy Serie A
5   13274       13274    Netherlands Eredivisie
6   15722       15722        Poland Ekstraklasa
7   17642       17642  Portugal Liga ZON Sagres
8   19694       19694   Scotland Premier League
9   21518       21518           Spain LIGA BBVA
10  24558       24558  Switzerland Super League

In [6]:
Country_table = pd.read_sql("""SELECT *
                        FROM Country;""", conn)
Country_table

id         name
0       1      Belgium
1    1729      England
2    4769       France
3    7809      Germany
4   10257        Italy
5   13274  Netherlands
6   15722       Poland
7   17642     Portugal
8   19694     Scotland
9   21518        Spain
10  24558  Switzerland

In [7]:
Team_table = pd.read_sql("""SELECT *
                        FROM Country;""", conn)
Team_table

id         name
0       1      Belgium
1    1729      England
2    4769       France
3    7809      Germany
4   10257        Italy
5   13274  Netherlands
6   15722       Poland
7   17642     Portugal
8   19694     Scotland
9   21518        Spain
10  24558  Switzerland

In [8]:
playerattributes_table = pd.read_sql("""SELECT *
                        FROM Player_Attributes;""", conn)
playerattributes_table

id  player_fifa_api_id  player_api_id                 date  \
0            1              218353         505942  2016-02-18 00:00:00   
1            2              218353         505942  2015-11-19 00:00:00   
2            3              218353         505942  2015-09-21 00:00:00   
3            4              218353         505942  2015-03-20 00:00:00   
4            5              218353         505942  2007-02-22 00:00:00   
...        ...                 ...            ...                  ...   
183973  183974              102359          39902  2009-08-30 00:00:00   
183974  183975              102359          39902  2009-02-22 00:00:00   
183975  183976              102359          39902  2008-08-30 00:00:00   
183976  183977              102359          39902  2007-08-30 00:00:00   
183977  183978              102359          39902  2007-02-22 00:00:00   

        overall_rating  potential preferred_foot attacking_work_rate  \
0                 67.0       71.0          right              medium   
1                 67.0       71.0          right              medium   
2                 62.0       66.0          right              medium   
3                 61.0       65.0          right              medium   
4                 61.0       65.0          right              medium   
...                ...        ...            ...                 ...   
183973            83.0       85.0          right              medium   
183974            78.0       80.0          right              medium   
183975            77.0       80.0          right              medium   
183976            78.0       81.0          right              medium   
183977            80.0       81.0          right              medium   

       defensive_work_rate  crossing  ...  vision  penalties  marking  \
0                   medium      49.0  ...    54.0       48.0     65.0   
1                   medium      49.0  ...    54.0       48.0     65.0   
2                   medium      49.0  ...    54.0       48.0     65.0   
3                   medium      48.0  ...    53.0       47.0     62.0   
4                   medium      48.0  ...    53.0       47.0     62.0   
...                    ...       ...  ...     ...        ...      ...   
183973                 low      84.0  ...    88.0       83.0     22.0   
183974                 low      74.0  ...    88.0       70.0     32.0   
183975                 low      74.0  ...    88.0       70.0     32.0   
183976                 low      74.0  ...    88.0       53.0     28.0   
183977                 low      74.0  ...    88.0       53.0     38.0   

        standing_tackle  sliding_tackle  gk_diving  gk_handling  gk_kicking  \
0                  69.0            69.0        6.0         11.0        10.0   
1                  69.0            69.0        6.0         11.0        10.0   
2                  66.0            69.0        6.0         11.0        10.0   
3                  63.0            66.0        5.0         10.0         9.0   
4                  63.0            66.0        5.0         10.0         9.0   
...                 ...             ...        ...          ...         ...   
183973             31.0            30.0        9.0         20.0        84.0   
183974             31.0            30.0        9.0         20.0        73.0   
183975             31.0            30.0        9.0         20.0        73.0   
183976             32.0            30.0        9.0         20.0        73.0   
183977             32.0            30.0        9.0          9.0        78.0   

        gk_positioning  gk_reflexes  
0                  8.0          8.0  
1                  8.0          8.0  
2                  8.0          8.0  
3                  7.0          7.0  
4                  7.0          7.0  
...                ...          ...  
183973            20.0         20.0  
183974            20.0         20.0  
183975            20.0         20.0  
183976            20.0         20.0  
183977        

In [9]:
teamattributes_table = pd.read_sql("""SELECT *
                        FROM Team_Attributes;""", conn)
teamattributes_table

id  team_fifa_api_id  team_api_id                 date  \
0        1               434         9930  2010-02-22 00:00:00   
1        2               434         9930  2014-09-19 00:00:00   
2        3               434         9930  2015-09-10 00:00:00   
3        4                77         8485  2010-02-22 00:00:00   
4        5                77         8485  2011-02-22 00:00:00   
...    ...               ...          ...                  ...   
1453  1454             15005        10000  2011-02-22 00:00:00   
1454  1455             15005        10000  2012-02-22 00:00:00   
1455  1456             15005        10000  2013-09-20 00:00:00   
1456  1457             15005        10000  2014-09-19 00:00:00   
1457  1458             15005        10000  2015-09-10 00:00:00   

      buildUpPlaySpeed buildUpPlaySpeedClass  buildUpPlayDribbling  \
0                   60              Balanced                   NaN   
1                   52              Balanced                  48.0   
2                   47              Balanced                  41.0   
3                   70                  Fast                   NaN   
4                   47              Balanced                   NaN   
...                ...                   ...                   ...   
1453                52              Balanced                   NaN   
1454                54              Balanced                   NaN   
1455                54              Balanced                   NaN   
1456                54              Balanced                  42.0   
1457                54              Balanced                  42.0   

     buildUpPlayDribblingClass  buildUpPlayPassing buildUpPlayPassingClass  \
0                       Little                  50                   Mixed   
1                       Normal                  56                   Mixed   
2                       Normal                  54                   Mixed   
3                       Little                  70                    Long   
4                       Little                  52                   Mixed   
...                        ...                 ...                     ...   
1453                    Little                  52                   Mixed   
1454                    Little                  51                   Mixed   
1455                    Little                  51                   Mixed   
1456                    Normal                  51                   Mixed   
1457                    Normal                  51                   Mixed   

      ... chanceCreationShooting  chanceCreationShootingClass  \
0     ...                     55                       Normal   
1     ...                     64                       Normal   
2     ...                     64                       Normal   
3     ...                     70                         Lots   
4     ...                     52                       Normal   
...   ...                    ...                          ...   
1453  ...                     53                       Normal   
1454  ...                     50                       Normal   
1455  ...                     32                       Little   
1456  ...                     32                       Little   
1457  ...                     32                       Little   

     chanceCreationPositioningClass  defencePressure defencePressureClass  \
0                         Organised               50               Medium   
1                         Organised               47               Medium   
2                         Organised               47               Medium   
3                         Organised               60               Medium   
4                         Organised               47               Medium   
...                             ...              ...                  ...   
1453                      Organised               46               Medium   
1454                      Organised     

Lets think as a football fan!

I want to take a look at which matches were played in a particular league and during a particular season and I probably want to take a look at scores as well.

To get the the desired results, we have to join match table with country table and then with league tabel then to the team table

Some of the fields in the different tables have same names. Renaming using AS in the query


In [10]:
detailed_matches = pd.read_sql("""
SELECT Match.id, 
Country.name AS country_name, 
League.name AS league_name, 
season, 
date,
HT.team_long_name AS  home_team,
AT.team_long_name AS away_team,
home_team_goal, 
away_team_goal                                        
FROM Match
LEFT JOIN Country on Country.id = Match.country_id
LEFT JOIN League on League.id = Match.league_id
LEFT JOIN Team AS HT on HT.team_api_id = Match.home_team_api_id
LEFT JOIN Team AS AT on AT.team_api_id = Match.away_team_api_id
WHERE country_name = 'England'
AND season = '2015/2016'
ORDER by date
LIMIT 10;""", conn)
detailed_matches

id country_name             league_name     season                 date  \
0  4390      England  England Premier League  2015/2016  2015-08-08 00:00:00   
1  4391      England  England Premier League  2015/2016  2015-08-08 00:00:00   
2  4392      England  England Premier League  2015/2016  2015-08-08 00:00:00   
3  4393      England  England Premier League  2015/2016  2015-08-08 00:00:00   
4  4394      England  England Premier League  2015/2016  2015-08-08 00:00:00   
5  4396      England  England Premier League  2015/2016  2015-08-08 00:00:00   
6  4389      England  England Premier League  2015/2016  2015-08-09 00:00:00   
7  4395      England  England Premier League  2015/2016  2015-08-09 00:00:00   
8  4397      England  England Premier League  2015/2016  2015-08-09 00:00:00   
9  4398      England  England Premier League  2015/2016  2015-08-10 00:00:00   

              home_team          away_team  home_team_goal  away_team_goal  
0           Bournemouth        Aston Villa               0               1  
1               Chelsea       Swansea City               2               2  
2               Everton            Watford               2               2  
3        Leicester City         Sunderland               4               2  
4     Manchester United  Tottenham Hotspur               1               0  
5          Norwich City     Crystal Palace               1               3  
6               Arsenal    West Ham United               0               2  
7      Newcastle United        Southampton               2               2  
8            Stoke City          Liverpool               0               1  
9  West Bromwich Albion    Manchester City               0               3

As a fan, i want to take a look at match results of the ream I follow.

In [11]:
tot = pd.read_sql("""
SELECT m.date,
HT.team_long_name AS  home_team,
AT.team_long_name AS away_team,
home_team_goal, 
away_team_goal, 
CASE WHEN m.home_team_api_id = 8586 and m.home_team_goal > m.away_team_goal THEN 'Spurs won'
WHEN m.home_team_api_id = 8586 and  m.home_team_goal < m.away_team_goal THEN 'Spurs lost'
WHEN m.away_team_api_id = 8586 and m.home_team_goal < m.away_team_goal THEN 'Spurs won'
WHEN m.away_team_api_id = 8586 and m.home_team_goal > m.away_team_goal THEN 'Spurs lost'
ELSE 'Tie' END AS outcome
FROM match m
LEFT JOIN Team AS HT on HT.team_api_id = m.home_team_api_id
LEFT JOIN Team AS AT on AT.team_api_id = m.away_team_api_id
WHERE m.home_team_api_id = 8586 or m.away_team_api_id = 8586
Order by 1;""", conn)
tot

date          home_team             away_team  \
0    2008-08-16 00:00:00      Middlesbrough     Tottenham Hotspur   
1    2008-08-23 00:00:00  Tottenham Hotspur            Sunderland   
2    2008-08-31 00:00:00            Chelsea     Tottenham Hotspur   
3    2008-09-15 00:00:00  Tottenham Hotspur           Aston Villa   
4    2008-09-21 00:00:00  Tottenham Hotspur        Wigan Athletic   
..                   ...                ...                   ...   
299  2016-04-18 00:00:00         Stoke City     Tottenham Hotspur   
300  2016-04-25 00:00:00  Tottenham Hotspur  West Bromwich Albion   
301  2016-05-02 00:00:00            Chelsea     Tottenham Hotspur   
302  2016-05-08 00:00:00  Tottenham Hotspur           Southampton   
303  2016-05-15 00:00:00   Newcastle United     Tottenham Hotspur   

     home_team_goal  away_team_goal     outcome  
0                 2               1  Spurs lost  
1                 1               2  Spurs lost  
2                 1               1         Tie  
3                 1               2  Spurs lost  
4                 0               0         Tie  
..              ...             ...         ...  
299               0               4   Spurs won  
300               1               1         Tie  
301               2               2         Tie  
302               1               2  Spurs lost  
303               5               1  Spurs lost  

[304 rows x 6 columns]

Now lets create a view for how my fav team performed each season.

In [12]:
totout = pd.read_sql("""
SELECT
m.season,
CASE WHEN m.home_team_api_id = 8586 and m.home_team_goal > m.away_team_goal THEN 'Spurs won'
WHEN m.home_team_api_id = 8586 and  m.home_team_goal < m.away_team_goal THEN 'Spurs lost'
WHEN m.away_team_api_id = 8586 and m.home_team_goal < m.away_team_goal THEN 'Spurs won'
WHEN m.away_team_api_id = 8586 and m.home_team_goal > m.away_team_goal THEN 'Spurs lost'
ELSE 'Tie' END AS outcome,
count (m.date) AS count
FROM match m
LEFT JOIN Team AS HT on HT.team_api_id = m.home_team_api_id
LEFT JOIN Team AS AT on AT.team_api_id = m.away_team_api_id
WHERE m.home_team_api_id = 8586 or m.away_team_api_id = 8586
GROUP BY m.season,outcome;""", conn)
totout

season     outcome  count
0   2008/2009  Spurs lost     15
1   2008/2009   Spurs won     14
2   2008/2009         Tie      9
3   2009/2010  Spurs lost     10
4   2009/2010   Spurs won     21
5   2009/2010         Tie      7
6   2010/2011  Spurs lost      8
7   2010/2011   Spurs won     16
8   2010/2011         Tie     14
9   2011/2012  Spurs lost      9
10  2011/2012   Spurs won     20
11  2011/2012         Tie      9
12  2012/2013  Spurs lost      8
13  2012/2013   Spurs won     21
14  2012/2013         Tie      9
15  2013/2014  Spurs lost     11
16  2013/2014   Spurs won     21
17  2013/2014         Tie      6
18  2014/2015  Spurs lost     12
19  2014/2015   Spurs won     19
20  2014/2015         Tie      7
21  2015/2016  Spurs lost      6
22  2015/2016   Spurs won     19
23  2015/2016         Tie     13

Lets create a view for how many goals a team scored across a season.

In [13]:
GoalsBySeason = pd.read_sql("""
SELECT
m.season,
SUM(CASE WHEN m.home_team_api_id = 8586  THEN home_team_goal END) AS home_goals,
SUM(CASE WHEN m.away_team_api_id = 8586  THEN away_team_goal END) AS away_goals
FROM match m
LEFT JOIN Team AS HT on HT.team_api_id = m.home_team_api_id
LEFT JOIN Team AS AT on AT.team_api_id = m.away_team_api_id
WHERE m.home_team_api_id = 8586 or m.away_team_api_id = 8586
GROUP BY m.season;""", conn)
GoalsBySeason

season  home_goals  away_goals
0  2008/2009          21          24
1  2009/2010          40          27
2  2010/2011          30          25
3  2011/2012          39          27
4  2012/2013          29          37
5  2013/2014          30          25
6  2014/2015          31          27
7  2015/2016          35          34

Lets make a view for average goals by season

In [14]:
AvgGoalsBySeason = pd.read_sql("""
SELECT
m.season,
ROUND(AVG(CASE WHEN m.home_team_api_id = 8586  THEN home_team_goal END),2) AS average_home_goals,
ROUND(AVG(CASE WHEN m.home_team_api_id = 8586  THEN away_team_goal END),2) AS average_away_goals
FROM match m
LEFT JOIN Team AS HT on HT.team_api_id = m.home_team_api_id
LEFT JOIN Team AS AT on AT.team_api_id = m.away_team_api_id
WHERE m.home_team_api_id = 8586 or m.away_team_api_id = 8586
GROUP BY m.season;""", conn)
AvgGoalsBySeason

season  average_home_goals  average_away_goals
0  2008/2009                1.11                0.53
1  2009/2010                2.11                0.63
2  2010/2011                1.58                1.00
3  2011/2012                2.05                0.89
4  2012/2013                1.53                0.95
5  2013/2014                1.58                1.21
6  2014/2015                1.63                1.26
7  2015/2016                1.84                0.79

In [15]:
AvgGoalsBySeason = pd.read_sql("""
SELECT
m.season,
ROUND(AVG(CASE WHEN m.home_team_api_id = 8586  THEN home_team_goal END),2) AS average_home_goals,
ROUND(AVG(CASE WHEN m.home_team_api_id = 8586  THEN away_team_goal END),2) AS average_away_goals
FROM match m
LEFT JOIN Team AS HT on HT.team_api_id = m.home_team_api_id
LEFT JOIN Team AS AT on AT.team_api_id = m.away_team_api_id
WHERE m.home_team_api_id = 8586 or m.away_team_api_id = 8586
GROUP BY m.season;""", conn)
AvgGoalsBySeason

season  average_home_goals  average_away_goals
0  2008/2009                1.11                0.53
1  2009/2010                2.11                0.63
2  2010/2011                1.58                1.00
3  2011/2012                2.05                0.89
4  2012/2013                1.53                0.95
5  2013/2014                1.58                1.21
6  2014/2015                1.63                1.26
7  2015/2016                1.84                0.79

Lets calculate a view for what percentage of home and away matches fav team won

In [16]:
AvgGoalsBySeasonPCT = pd.read_sql(
"""
SELECT
m.season,
ROUND(AVG(CASE WHEN m.home_team_api_id = 8586 AND home_team_goal > away_team_goal THEN 1
WHEN m.home_team_api_id = 8586 AND home_team_goal < away_team_goal THEN 0 END),2) AS  pct_homewins,
ROUND(AVG(CASE WHEN m.away_team_api_id = 8586 AND away_team_goal > home_team_goal THEN 1
WHEN m.home_team_api_id = 8586 AND away_team_goal < home_team_goal THEN 0 END),2) AS  pct_awaywins
FROM match m 
LEFT JOIN Team AS HT on HT.team_api_id = m.home_team_api_id
LEFT JOIN Team AS AT on AT.team_api_id = m.away_team_api_id
WHERE m.home_team_api_id = 8586 or m.away_team_api_id = 8586
GROUP BY m.season;""", conn)
AvgGoalsBySeasonPCT

season  pct_homewins  pct_awaywins
0  2008/2009          0.71          0.29
1  2009/2010          0.82          0.33
2  2010/2011          0.90          0.44
3  2011/2012          0.81          0.35
4  2012/2013          0.79          0.48
5  2013/2014          0.69          0.48
6  2014/2015          0.63          0.47
7  2015/2016          0.77          0.47

Lets create a view for what percentage of matches end in draw across different countries and leagues

To keep the comparison focussed lets just compare seasons 2013/14 and 2014/15.



In [17]:
perdraw = pd.read_sql("""
SELECT c.name AS country,
ROUND(AVG(CASE WHEN m.season='2013/2014' AND m.home_team_goal = m.away_team_goal THEN 1
 WHEN m.season='2013/2014' AND m.home_team_goal != m.away_team_goal THEN 0
END),2) AS pct_ties_2013_2014,
ROUND(AVG(CASE WHEN m.season='2014/2015' AND m.home_team_goal = m.away_team_goal THEN 1
 WHEN m.season='2014/2015' AND m.home_team_goal != m.away_team_goal THEN 0
END),2) AS pct_ties_2014_2015
FROM country AS c
LEFT JOIN match AS m
ON c.id = m.country_id
GROUP BY country
;""", conn)
perdraw

country  pct_ties_2013_2014  pct_ties_2014_2015
0       Belgium                0.17                0.25
1       England                0.21                0.24
2        France                0.28                0.23
3       Germany                0.21                0.27
4         Italy                0.24                0.32
5   Netherlands                0.27                0.24
6        Poland                0.30                0.28
7      Portugal                0.25                0.28
8      Scotland                0.22                0.19
9         Spain                0.23                0.24
10  Switzerland                0.23                0.27

In the 2014/2015 season, Italy had the highest proportion of draws (almost 1/3) and Scotland had the lowest (about 1/5)

Following on from this, what league had the most number of goals, averaging out across all the matches within that league?

In [18]:
leagues_by_avg_goal = pd.read_sql("""
SELECT 
l.name AS league,
ROUND(AVG(m.home_team_goal + m.away_team_goal),2) AS avg_goals
FROM league as L
LEFT JOIN match AS m
ON l.id = m.country_id
WHERE m.season = '2014/2015'
GROUP BY league
;""", conn)
leagues_by_avg_goal

league  avg_goals
0     Belgium Jupiler League       2.78
1     England Premier League       2.57
2             France Ligue 1       2.49
3      Germany 1. Bundesliga       2.75
4              Italy Serie A       2.69
5     Netherlands Eredivisie       3.08
6         Poland Ekstraklasa       2.62
7   Portugal Liga ZON Sagres       2.49
8    Scotland Premier League       2.57
9            Spain LIGA BBVA       2.66
10  Switzerland Super League       2.87

We can see that the Netherlands had the highest average number of goals with 3.08.

 what was the average number of goals,across all leagues, during the 2014/15 season.

In [19]:
avg_goal1415 = pd.read_sql("""
SELECT 
ROUND(AVG(m.home_team_goal + m.away_team_goal),2) AS avg_goals
FROM match AS m
WHERE m.season = '2014/2015';""", conn)
avg_goal1415

avg_goals
0       2.68

Number of matches in the French league, where more goals were scored than this average.

In [20]:
Higher_than_avg_goals1415 = pd.read_sql("""
SELECT 
HT.team_long_name AS  home_team,
AT.team_long_name AS away_team,
m.home_team_goal,
m.away_team_goal
FROM match AS m
LEFT JOIN League as l
ON l.id = m.country_id
LEFT JOIN Team AS HT on HT.team_api_id = m.home_team_api_id
LEFT JOIN Team AS AT on AT.team_api_id = m.away_team_api_id
WHERE m.season = '2014/2015'
AND l.name = 'France Ligue 1'
AND (m.home_team_goal + m.away_team_goal) >
(SELECT 
    AVG(m.home_team_goal + m.away_team_goal) AS avg_goals
    FROM match AS m
WHERE m.season = '2014/2015');""", conn)
Higher_than_avg_goals1415

home_team               away_team  home_team_goal  \
0              Stade de Reims     Paris Saint-Germain               2   
1                   SC Bastia  Olympique de Marseille               3   
2    Évian Thonon Gaillard FC                 SM Caen               0   
3                   AS Monaco              FC Lorient               1   
4                    OGC Nice             Toulouse FC               3   
..                        ...                     ...             ...   
165                   FC Metz          Stade de Reims               3   
166    Montpellier Hérault SC    En Avant de Guingamp               2   
167                   SM Caen  Olympique de Marseille               1   
168  Évian Thonon Gaillard FC                 FC Metz               3   
169        Olympique Lyonnais              LOSC Lille               3   

     away_team_goal  
0                 2  
1                 3  
2                 3  
3                 2  
4                 2  
..              ...  
165               0  
166               1  
167               2  
168               0  
169               0  

[170 rows x 4 columns]

lets pull all the matches in the database, but then only those matches where more than 10 goals were scored.

In [21]:
matchesabove10 = pd.read_sql("""
SELECT 
country,date,total_goals
FROM 
(SELECT 
c.name as country,
m.date,
m.home_team_goal,
m.away_team_goal,
(m.home_team_goal + m.away_team_goal) as total_goals
from match AS m
left join country as c
ON m.country_id = c.id) AS subquery
WHERE 
total_goals >=10
    ;""", conn)
matchesabove10

country                 date  total_goals
0       England  2009-11-22 00:00:00           10
1       England  2011-08-28 00:00:00           10
2       England  2012-12-29 00:00:00           10
3       England  2013-05-19 00:00:00           10
4        France  2009-11-08 00:00:00           10
5       Germany  2013-03-30 00:00:00           11
6   Netherlands  2010-10-24 00:00:00           10
7   Netherlands  2011-11-06 00:00:00           10
8      Scotland  2010-05-05 00:00:00           12
9         Spain  2013-10-30 00:00:00           10
10        Spain  2015-04-05 00:00:00           10
11        Spain  2015-05-23 00:00:00           10
12        Spain  2014-09-20 00:00:00           10
13        Spain  2015-12-20 00:00:00           12

by using a subquery within the SELECT clause, the total goals scored in each 2014/15 match can be benchmarked against the average goals per match across all seasons. This demonstrates using subqueries to derive comparison metrics within a result set.

In [22]:
TotalGoalsVAvG = pd.read_sql("""
SELECT 
date,
(m.home_team_goal + m.away_team_goal) AS total_goals,
(SELECT ROUND(AVG(m.home_team_goal + m.away_team_goal),2)
FROM match AS m
WHERE season = '2014/2015') AS overall_avg
FROM match m
WHERE m.season = '2014/2015'
    ;""", conn)
TotalGoalsVAvG

date  total_goals  overall_avg
0     2014-07-25 00:00:00            3         2.68
1     2014-07-26 00:00:00            2         2.68
2     2014-07-26 00:00:00            0         2.68
3     2014-07-26 00:00:00            2         2.68
4     2014-07-26 00:00:00            1         2.68
...                   ...          ...          ...
3320  2014-09-23 00:00:00            4         2.68
3321  2014-09-23 00:00:00            5         2.68
3322  2014-09-24 00:00:00            3         2.68
3323  2014-09-24 00:00:00            0         2.68
3324  2014-09-24 00:00:00            1         2.68

[3325 rows x 3 columns]

Lets calculate the average total of goals scored within each league

Lets calculate avg total of goals scored in each league and also use a sub query in the select statement to subtract the overall average from each leagues' average. Through this we get a sense of where each league stands in relation to the average.

In [23]:
goalsdiff = pd.read_sql("""
Select 
l.name as league,
round((avg(m.home_team_goal + m.away_team_goal)),2) as avg_goals,
round(avg(m.home_team_goal + m.away_team_goal)
-((Select avg(home_team_goal + away_team_goal) from match WHERE season = '2014/2015')),2) as diff 
FROM league as l
LEFT JOIN match as m
ON l.country_id = m.country_id
WHERE m.season = '2014/2015'
Group by l.name
    ;""", conn)
goalsdiff

league  avg_goals  diff
0     Belgium Jupiler League       2.78  0.11
1     England Premier League       2.57 -0.11
2             France Ligue 1       2.49 -0.18
3      Germany 1. Bundesliga       2.75  0.08
4              Italy Serie A       2.69  0.01
5     Netherlands Eredivisie       3.08  0.40
6         Poland Ekstraklasa       2.62 -0.06
7   Portugal Liga ZON Sagres       2.49 -0.18
8    Scotland Premier League       2.57 -0.10
9            Spain LIGA BBVA       2.66 -0.02
10  Switzerland Super League       2.87  0.20

lets create a CTE which pulls all the matches where 10 or more goals were scored. Then lets join that information onto the league table, so we get a count of how matches have occurred in each league

In [24]:
HighScoreMatchesByLeague = pd.read_sql("""
With big_game AS (
Select 
league_id
,id
from match 
WHERE home_team_goal + away_team_goal >= 10)
Select
l.name as league,
COUNT(big_game.id) as High_Score_Matches
FROM league as l
LEFT JOIN big_game
ON l.id = big_game.league_id
GROUP BY league
    ;""", conn)
HighScoreMatchesByLeague

league  High_Score_Matches
0     Belgium Jupiler League                   0
1     England Premier League                   4
2             France Ligue 1                   1
3      Germany 1. Bundesliga                   1
4              Italy Serie A                   0
5     Netherlands Eredivisie                   2
6         Poland Ekstraklasa                   0
7   Portugal Liga ZON Sagres                   0
8    Scotland Premier League                   1
9            Spain LIGA BBVA                   5
10  Switzerland Super League                   0

There are leagues who never had a single match where 10 or more goals were scored. At the other extreme, spains LIGA BBVA had the highest number of matches where this happened.

Lets create a query that sums up the total number of goals that were scored in matches and then joining that information to an outer query that sums up the information by a dimension only available outside of CTE, lets say league name.

In [25]:
AvgGoalsByleague = pd.read_sql("""
With match_list as
(
Select 
country_id
,(home_team_goal + away_team_goal) AS goals
from match
)
Select name
,ROUND(AVG (goals),2) as Average_Goals
FROM league
LEFT JOIN match_list 
ON league.id = match_list.country_id
Group by name

 ;""", conn)

AvgGoalsByleague

name  Average_Goals
0     Belgium Jupiler League           2.80
1     England Premier League           2.71
2             France Ligue 1           2.44
3      Germany 1. Bundesliga           2.90
4              Italy Serie A           2.62
5     Netherlands Eredivisie           3.08
6         Poland Ekstraklasa           2.43
7   Portugal Liga ZON Sagres           2.53
8    Scotland Premier League           2.63
9            Spain LIGA BBVA           2.77
10  Switzerland Super League           2.93